napari/components/add_layers_mixin.py

```diff
@@ -58,7 +58,11 @@ def add_layer(self, layer: layers.Layer) -> layers.Layer:
         layer.events.cursor.connect(self._update_cursor)
         layer.events.cursor_size.connect(self._update_cursor_size)
         layer.events.data.connect(self._on_layers_change)
+        layer.name = self.layers._coerce_name(layer.name, layer)
```

Comment:
I just want to note one design flaw we have right now which this PR isn't really making worse but just exposing further - if you call `viewer.layers.append(layer)` instead of `viewer.add_layer(layer)` than none of this wiring up happens correctly, and now the name coercing won't happen either. Not sure if we should try and address this now @tlambert03 @jni?

The ways I can see it are we pull out all the event connection into its own method that gets triggered via events either when calling `add_layer` or `layers.append`, but this logic of events triggering the connection of events is really ugly to maintain, especially in this case where we have to deal with `layers.inserted` which gets emitted during a layer move too, or we work to eliminate the need for all these events, but that will be hard

Reply:
yeah, that `AddLayersMixin.add_layer` method has been something that I've wanted to see disappear for a while.  I would like to see a situation where the viewer only listens to a `LayerList`, and simply inserting/removing layers from that list is sufficient. (that's going to be important if the `LayerList` is to eventually be our "unit" of a group of layers... layers shouldn't be connected both to the viewer and the LayerList, but rather layer to List and List to viewer)

If you've been able to get everything working here as it was before, then I think all that should be addressed in a later PR (since it's not a "new" problem)

Reply:
Ok nice, we see this the same way. The one change this PR would make right now is the coercing of uniqueness in the name would only happen on `viewer.layers` and making only the last layer selected would only happen on `viewer.layers`, but otherwise as you say this is not a new problem. If we can live with that then I'll make a new issue right now describing this and we can merge this PR without anything more here

Reply:
issue made here https://github.com/napari/napari/issues/1846